numpy/ma/core.py

```diff
@@ -2726,7 +2726,7 @@ def __getitem__(self, indx):
             _mask = self.maskiter.__getitem__(indx)
             if isinstance(_mask, ndarray):
                 # set shape to match that of data; this is needed for matrices
-                _mask.shape = result.shape
+                _mask = _mask.reshape(result.shape)
                 result._mask = _mask
```

Comment:
Jut put the reshape here, `result._mask = _mask.reshape(result.shape)` (like you do just below).